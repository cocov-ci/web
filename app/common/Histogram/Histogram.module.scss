@use "sass:map";
@import "styles/variables.module";

.base {
  display: flex;

  .labels {
    text-align: right;

    .label {
      display: flex;
      align-items: center;
      justify-content: flex-end;
      height: 32px;
      margin-bottom: 12px;

      .labelWrapper {
        padding: 8px 12px;
      }
    }
  }

  .values {
    flex: 1;

    .value {
      height: 32px;
      margin-bottom: 12px;
      margin-left: 20px;

      .valueWrapper {
        padding: 8px 12px;
        font-variation-settings: "wght" 500;
        font-size: 12px;
        color: map.get($colorHistogram, "value-color");
        letter-spacing: 0.96px;
        background-image:
          linear-gradient(
            180deg,
            map.get($colorHistogram, "value-background-1") 0%,
            map.get($colorHistogram, "value-background-2") 100%
          );
        border: 1px solid #a5d1f4;
        border-radius: 8px;
      }
    }
  }
}
