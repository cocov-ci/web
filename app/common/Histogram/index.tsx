import classNames from 'classnames'
import Link from 'next/link'
import React, { HTMLAttributeAnchorTarget } from 'react'

import styles from './Histogram.module.scss'

type HistogramItem = {
  label: string
  value: number
  href?: string
  hrefTarget?: HTMLAttributeAnchorTarget
}

type HistogramProps = {
  className?: string
  values: Array<HistogramItem>
}

const Histogram = ({ className, values }: HistogramProps) => {
  const maxValue = Math.max(0.01, ...values.map(i => i.value))

  const autoWrap = (el: HistogramItem, children: JSX.Element) => {
    if (el.href) {
      return (
        <Link href={el.href} key={`label-${el.label}`} target={el.hrefTarget}>
          {children}
        </Link>
      )
    }

    return children
  }

  return (
    <div className={classNames(styles.base, className)}>
      <div className={styles.labels}>
        {values.map(obj =>
          autoWrap(
            obj,
            <div className={styles.label} key={`label-${obj.label}`}>
              <div className={styles.labelWrapper}>{obj.label}</div>
            </div>,
          ),
        )}
      </div>
      <div className={styles.values}>
        {values.map(obj =>
          autoWrap(
            obj,
            <div className={styles.value} key={`value-${obj.label}`}>
              <div
                className={styles.valueWrapper}
                style={{
                  width: `${(obj.value / maxValue) * 95}%`,
                }}
              >
                {obj.value}
              </div>
            </div>,
          ),
        )}
      </div>
    </div>
  )
}

export default Histogram
