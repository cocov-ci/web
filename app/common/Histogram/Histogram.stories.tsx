import type { Meta, StoryObj } from '@storybook/react'

import Histogram from './index'

const meta: Meta<typeof Histogram> = {
  title: 'Common/Histogram',
  component: Histogram,
  tags: ['autodocs'],
}

export default meta

type Story = StoryObj<typeof Histogram>

export const Default: Story = {
  args: {
    values: [
      'Zero',
      'Ten',
      'Twenty',
      'Thirty',
      'Forty',
      'Fifty',
      'Sixty',
      'Seventy',
      'Eighty',
      'Ninety',
      'A Hundred',
    ].map((i, idx) => ({
      value: idx * 10,
      label: i,
      href: '/',
    })),
  },
}
