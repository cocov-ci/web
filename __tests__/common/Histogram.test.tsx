import { render, screen } from '@testing-library/react'

import Histogram from 'app/common/Histogram'

import '@testing-library/jest-dom'

const args = {
  values: [
    'Zero',
    'Ten',
    'Twenty',
    'Thirty',
    'Forty',
    'Fifty',
    'Sixty',
    'Seventy',
    'Eighty',
    'Ninety',
    'A Hundred',
  ].map((i, idx) => ({
    value: idx * 10,
    label: i,
    href: '/',
  })),
}

describe('common/Histogram', () => {
  it('renders Histogram', () => {
    render(<Histogram {...args} />)

    expect(screen.getByText('Zero')).toBeVisible()
    expect(screen.getByText('0')).toBeVisible()
    expect(screen.getByText('Ten')).toBeVisible()
    expect(screen.getByText('10')).toBeVisible()
  })

  it('renders Histogram without link', () => {
    const args = {
      values: ['Zero'].map((i, idx) => ({
        value: idx * 10,
        label: i,
      })),
    }

    render(<Histogram {...args} />)

    expect(screen.getByText('Zero')).toBeVisible()
    expect(screen.getByText('0')).toBeVisible()
  })

  it('renders Histogram snapshots', () => {
    const { container } = render(<Histogram {...args} />)

    expect(container).toMatchSnapshot()
  })
})
