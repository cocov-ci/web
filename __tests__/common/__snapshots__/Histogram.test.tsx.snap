// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Histogram renders Histogram snapshots 1`] = `
<div>
  <div
    class="base"
  >
    <div
      class="labels"
    >
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Zero
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Ten
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Twenty
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Thirty
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Forty
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Fifty
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Sixty
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Seventy
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Eighty
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            Ninety
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="label"
        >
          <div
            class="labelWrapper"
          >
            A Hundred
          </div>
        </div>
      </a>
    </div>
    <div
      class="values"
    >
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 0%;"
          >
            0
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 9.5%;"
          >
            10
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 19%;"
          >
            20
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 28.5%;"
          >
            30
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 38%;"
          >
            40
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 47.5%;"
          >
            50
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 57%;"
          >
            60
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 66.5%;"
          >
            70
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 76%;"
          >
            80
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 85.5%;"
          >
            90
          </div>
        </div>
      </a>
      <a
        href="/"
      >
        <div
          class="value"
        >
          <div
            class="valueWrapper"
            style="width: 95%;"
          >
            100
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
